### PR TITLE
Add IOP CreateThread test

### PR DESCRIPTION
**Pull request by jpd002**
Basic test cases for IOP's CreateThread function.

One weird thing is that I wasn't able to use the `TH_NO_FILLSTACK` and `TH_CLEAR_STACK` attribute bits even though they are defined in [ps2sdk](https://github.com/ps2dev/ps2sdk/blob/747b590ff1136ad000a4dbbc733dfd41ce7dcdc8/iop/kernel/include/thbase.h#L24). I think it's because I'm running the tests on one of the early PS2 models which might have a very old version of the BIOS.

One thing I'm thinking of doing at some point is to modify ps2client to allow loading of a specific version of an IOPRP image to simulate different kernel versions. Maybe it would be interesting to have `expected` files attached to kernel module versions to document the important changes between kernel iterations.

### CODE REVIEW

**Comment by unknownbrackets** [comment]
Hmm, that sounds likely.  I've heard that some PS2 games "hotpatch" the bios (possibly even custom in some cases?)

In pspautotests, we have a switch to test.py that runs the test against each kernel version, and reports which (if any) returned different results.  We've used that to contruct tests for that functionality.  In the PSP's case, this is fairly easy, because there's a group of syscalls used to select the kernel version.

I'd want to avoid having a separate expected file for each version of the ee cpu tests, but for HLE, it may make sense....

-[Unknown]

File: tests/kernel/iop/thread/create.c [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	createThread(0, &nullThreadProc, 10, 300, 0);
+	
+	printf("invalid attr (~0):\n");
+	createThread(~0, &nullThreadProc, 10, 0x800, 0);
+	
+	//TH_NO_FILLSTACK and TH_CLEAR_STACK are defined in ps2sdk, but don't seem to be supported
+	//(Might be due to BIOS version differences)
+	
+	printf("known attribute (TH_NO_FILLSTACK):\n");
+	createThread(TH_NO_FILLSTACK, &nullThreadProc, 10, 0x800, 0);
```
Comment (unknownbrackets) [inline tests/kernel/iop/thread/create.c:67]: Did you try pairing with `TH_C` or etc.?  Maybe they are only valid in certain combinations?

-[Unknown]

File: tests/kernel/iop/thread/create.c [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	} else {
+		printf("  failed -> result: %d\n", thId);
+	}
+}
+
+int _start(int argc, char **argv) {
+	printf("-- TEST BEGIN\n");
+	
+	printf("valid:\n");
+	createThread(TH_C, &nullThreadProc, 10, 0x800, 0x01234567);
```
Comment (unknownbrackets) [inline tests/kernel/iop/thread/create.c:52]: If I create a thread, delete it, and then create another thread, will the stack be positioned in the same place?  Or does it use a rolling allocation algorithm?

-[Unknown]

File: tests/kernel/iop/thread/create.c [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	
+	printf("huge stack size (~0):\n");
+	createThread(0, &nullThreadProc, 10, ~0, 0);
+	
+	printf("small stack size (300):\n");
+	createThread(0, &nullThreadProc, 10, 300, 0);
```
Comment (unknownbrackets) [inline tests/kernel/iop/thread/create.c:58]: Did you try alignment?  I think the PSP requires the stack to be aligned to 16, which is really a MIPS requirement.  Both 300 and ~0 happen to not be aligned, which is why I ask.

-[Unknown]

File: tests/kernel/iop/thread/create.expected [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  failed -> result: -401
+known attribute (TH_CLEAR_STACK):
+  failed -> result: -401
+known valid attributes (TH_UMODE | TH_ASM | TH_C):
+  succeeded -> attr: 03000008, option: 00000000, stack size: 00000800, init prio: 0a, curr prio: 00, status: 10, entry: &nullThreadProc
+invalid entry point (NULL):
+  succeeded -> attr: 00000000, option: 00000000, stack size: 00000800, init prio: 0a, curr prio: 00, status: 10, entry: 00000000
+invalid entry point (3):
+  failed -> result: -402
+invalid entry point (~3):
```
Comment (unknownbrackets) [inline tests/kernel/iop/thread/create.expected:20]: Could this be an alignment check too?  ~3 happens to be aligned.

-[Unknown]

File: tests/kernel/iop/thread/create.c [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	printf("invalid entry point (~3):\n");
+	createThread(0, (void*)~3, 10, 0x800, 0);
+	
+	printf("invalid priority (0):\n");
+	createThread(0, &nullThreadProc, 0, 0x800, 0);
+	
+	printf("invalid priority (127):\n");
+	createThread(0, &nullThreadProc, 127, 0x800, 0);
+	
+	printf("-- TEST END\n");
```
Comment (unknownbrackets) [inline tests/kernel/iop/thread/create.c:90]: Another thing to test might be total # of threads.  If you create in a loop, does it fail after some number?

-[Unknown]

File: tests/kernel/iop/thread/create.c [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	createThread(0, (void*)3, 10, 0x800, 0);
+	
+	printf("invalid entry point (~3):\n");
+	createThread(0, (void*)~3, 10, 0x800, 0);
+	
+	printf("invalid priority (0):\n");
+	createThread(0, &nullThreadProc, 0, 0x800, 0);
+	
+	printf("invalid priority (127):\n");
+	createThread(0, &nullThreadProc, 127, 0x800, 0);
```
Comment (unknownbrackets) [inline tests/kernel/iop/thread/create.c:88]: Is the valid priority different for different types of threads (by attr)?  Does TH_UMODE affect it?

-[Unknown]

File: tests/kernel/iop/thread/create.c [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	//TH_NO_FILLSTACK and TH_CLEAR_STACK are defined in ps2sdk, but don't seem to be supported
+	//(Might be due to BIOS version differences)
+	
+	printf("known attribute (TH_NO_FILLSTACK):\n");
+	createThread(TH_NO_FILLSTACK, &nullThreadProc, 10, 0x800, 0);
+	
+	printf("known attribute (TH_CLEAR_STACK):\n");
+	createThread(TH_CLEAR_STACK, &nullThreadProc, 10, 0x800, 0);
+	
+	printf("known valid attributes (TH_UMODE | TH_ASM | TH_C):\n");
```
Comment (unknownbrackets) [inline tests/kernel/iop/thread/create.c:72]: A pcsx2 file references 0x404 and TH_COP:
https://github.com/PCSX2/pcsx2/blob/master/unfree/fps2bios/kernel/iopload/threadman/threadman.c#L511

Not sure what all it means, I wonder if it'd be interesting to test those bits.

-[Unknown]

**Comment by jpd002** [comment]
Well, from what I know, almost all games will reboot the IOP and tell it to load an IOPRP image which contains updated versions of IOP kernel modules (thbase, thsema, timrman, etc.). AFAIK, there's no such mechanism on the EE. I've seen some games using custom system call handlers, but I've never seen a game updating the BIOS' code.

And yeah, separate expected files would only make sense for features that might be emulated using HLE and it would probably be interesting to only keep files that show some difference with previous versions.

**Comment by unknownbrackets** [comment]
Might be I misunderstood - that definitely sounds a lot more sane and workable.  Although custom syscall handlers sounds annoying...

-[Unknown]

File: tests/kernel/iop/thread/create.c (hunk 7 above)
Comment (jpd002) [inline tests/kernel/iop/thread/create.c:72]: Interesting indeed. I don't see `TH_COP` defined anywhere, but there might be a few bits that are not defined but that are still allowable: https://github.com/PCSX2/pcsx2/blob/master/unfree/fps2bios/kernel/iopload/threadman/threadman.c#L430.

File: tests/kernel/iop/thread/create.expected (hunk 4 above)
Comment (jpd002) [inline tests/kernel/iop/thread/create.expected:20]: Yeah, my idea here was to have an address that's way out of range but still aligned to see if CreateThread is only checking alignment to determine if an entry point address is valid.

**Comment by unknownbrackets** [comment]
Merging - this is a pretty useful test even without the couple of other ideas I threw out.

-[Unknown]

File: tests/kernel/iop/thread/create.expected (hunk 4 above)
Comment (unknownbrackets) [inline tests/kernel/iop/thread/create.expected:20]: Hmm, I would normally use 0xFFFFFFFC to indicate I'm using an aligned value specifically - I mean, I know ~3 is one, but it's "burying the lead" if that's the main point of the number... like a double negative.  But doesn't matter.

-[Unknown]

Comment (jpd002) [inline tests/kernel/iop/thread/create.expected:20]: Oh yeah, gotcha. I'll avoid using "tricks" like those to generate numbers in the future.

**Comment by jpd002** [comment]
Thanks. I'll probably revisit the test once I find a game that uses those stack fill/clean flags to see how they are handled by the kernel version used by the game.

**Pull request merged**